src/mscorlib/src/System/Runtime/CompilerServices/ConditionalWeakTable.cs

```diff
@@ -150,6 +150,41 @@ public void Add(TKey key, TValue value)
         }
 
         //--------------------------------------------------------------------------------------------
+        // key: key to add or update. May not be null.
+        // value: value to associate with key.
+        //
+        // If the key is already entered into the dictionary, this method will update the value associated with key.
+        //--------------------------------------------------------------------------------------------
+        [System.Security.SecurityCritical]
+        public void AddOrUpdate(TKey key, TValue value)
+        {
+            if (key == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.key);
+            }
+
+            lock (_lock)
+            {
+                VerifyIntegrity();
+                _invalid = true;
+                int entryIndex = FindEntry(key);
+
+                // if we found a key we should just update, if no we should create a new entry.
+                if (entryIndex != -1)
+                {
+                    _entries[entryIndex].depHnd.Free();
```

Comment:
It is waste of cycles to free the handle and create a new one. Instead, you should just call `SetDependentHandleSecondary` on the existing handle. (You will need to add a new FCall for SetDependentHandleSecondary.)